waku/waku_store_v3/protocol.nim

```diff
@@ -0,0 +1,107 @@
+## Waku Store protocol for historical messaging support.
+## See spec for more details:
+## https://github.com/vacp2p/specs/blob/master/specs/waku/v2/waku-store.md
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import
+  std/options,
+  stew/results,
+  chronicles,
+  chronos,
+  bearssl/rand,
+  libp2p/crypto/crypto,
+  libp2p/protocols/protocol,
+  libp2p/protobuf/minprotobuf,
+  libp2p/stream/connection,
+  metrics
+import
+  ../waku_core,
+  ../node/peer_manager,
+  ./common,
+  ./rpc_codec,
+  ./protocol_metrics
+
+logScope:
+  topics = "waku store"
+
+const
+  MaxMessageTimestampVariance* = getNanoSecondTime(20) # 20 seconds maximum allowable sender timestamp "drift"
+
+type StoreRequestHandler* = proc(req: StoreRequest): Future[StoreResult] {.async, gcsafe.}
+
+type
+  WakuStoreV3* = ref object of LPProtocol
+    peerManager: PeerManager
+    rng: ref rand.HmacDrbgContext
+    requestHandler*: StoreRequestHandler
+
+## Protocol
+
+proc initProtocolHandler(self: WakuStoreV3) =
+
+  proc handler(conn: Connection, proto: string) {.async.} =
+    let buf = await conn.readLp(MaxRpcSize.int)
+
+    var res = StoreResponse()
+
+    let req = StoreRequest.decode(buf).valueOr:
+      error "failed to decode rpc", peerId= $conn.peerId
+      waku_store_errors.inc(labelValues = [decodeRpcFailure])
+      
+      res.statusCode = some(uint32(400))# TODO find a way to use the error enum
+      res.statusDesc = some("decode rpc failed")
+
+      await conn.writeLp(res.encode().buffer)
+      return
+
```

Comment:
After decoding, I would pull out `handleStoreRequest` into a separate `proc` (for unit testing and better separation of interests).